transport/http/src/test/java/org/kaazing/gateway/transport/http/security/auth/HttpAcceptorAuthenticationIT.java

```diff
@@ -0,0 +1,305 @@
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.kaazing.gateway.transport.http.security.auth;
+
+import static java.util.concurrent.TimeUnit.SECONDS;
+
+import javax.security.auth.Subject;
+import javax.security.auth.login.LoginException;
+
+import org.jmock.Expectations;
+import org.jmock.integration.junit4.JUnitRuleMockery;
+import org.jmock.lib.concurrent.Synchroniser;
+import org.jmock.lib.legacy.ClassImposteriser;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.DisableOnDebug;
+import org.junit.rules.RuleChain;
+import org.junit.rules.TestRule;
+import org.junit.rules.Timeout;
+import org.kaazing.gateway.resource.address.http.DefaultHttpRealmInfo;
+import org.kaazing.gateway.resource.address.http.HttpRealmInfo;
+import org.kaazing.gateway.security.LoginContextFactory;
+import org.kaazing.gateway.security.TypedCallbackHandlerMap;
+import org.kaazing.gateway.security.auth.DefaultLoginResult;
+import org.kaazing.gateway.security.auth.context.DefaultLoginContextFactory;
+import org.kaazing.gateway.security.auth.context.ResultAwareLoginContext;
+import org.kaazing.gateway.server.spi.security.LoginResult;
+import org.kaazing.gateway.transport.IoHandlerAdapter;
+import org.kaazing.gateway.transport.http.HttpAcceptorRule;
+import org.kaazing.k3po.junit.annotation.Specification;
+import org.kaazing.k3po.junit.rules.K3poRule;
+import org.kaazing.test.util.MethodExecutionTrace;
+
+public class HttpAcceptorAuthenticationIT {
+
+    private static final String REALM_NAME = "Test Realm";
+    private static final String BIND_ADDRESS = "http://localhost:8000/resource";
+    private static final String BASIC_CHALLENGE_SCHEME = "Basic";
+    private static final Object[] ADDITIONAL_CHALLENGES = new Object[]{"param=\"value\""};
+
+    private final HttpAcceptorRule acceptor = new HttpAcceptorRule();
+
+    private final TestRule trace = new MethodExecutionTrace();
+    private final K3poRule k3po = new K3poRule();
+    private final TestRule timeoutRule = new DisableOnDebug(new Timeout(5, SECONDS));
+
+    @Rule
+    public JUnitRuleMockery context = new JUnitRuleMockery() {
+        {
+            setImposteriser(ClassImposteriser.INSTANCE);
+            setThreadingPolicy(new Synchroniser());
+        }
+    };
+
+    @Rule
+    public TestRule chain = RuleChain.outerRule(trace).around(acceptor).around(k3po).around(timeoutRule);
+
+    @Test
+    @Specification("client.no.credentials.unauthorized.empty.additional.challenges")
+    public void testGatewayReturns401WithEmptyAdditionalChallengeDataWhenNoAuthenticationCredentialsProvided() throws Exception {
```

Comment:
Excellent test, but the title of this test, specifically the "WhenNoAuthenticationCredentialsProvided" makes it seem like you are testing both the HttpSecurityFilter behavior and the LoginModule behavior.  When in fact you really just want to test the HttpSecurityFilter behavior here for all events it might see from the login modules.  I would suggest renaming the methods to something of the form, return401WhenLoginModuleThrowsLoginExceptionAndSetChallenge (or related per test).